File: packages/@juspay-tech/react-native-hyperswitch-klarna-expresscheckout/android/src/main/java/com/juspaytech/reactnativehyperswitchklarnaexpresscheckout/KlarnaExpressCheckoutButtonView.kt
package com.juspaytech.reactnativehyperswitchklarnaexpresscheckout

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.view.ViewGroup
import android.widget.FrameLayout
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.events.RCTEventEmitter
import com.klarna.mobile.sdk.api.KlarnaEnvironment
import com.klarna.mobile.sdk.api.KlarnaRegion
import com.klarna.mobile.sdk.api.KlarnaTheme
import com.klarna.mobile.sdk.api.KlarnaLoggingLevel
import com.klarna.mobile.sdk.api.checkout.express.KlarnaExpressCheckoutButton
import com.klarna.mobile.sdk.api.checkout.express.KlarnaExpressCheckoutButtonAuthorizationResponse
import com.klarna.mobile.sdk.api.checkout.express.KlarnaExpressCheckoutButtonCallback
import com.klarna.mobile.sdk.api.checkout.express.KlarnaExpressCheckoutButtonOptions
import com.klarna.mobile.sdk.api.checkout.express.KlarnaExpressCheckoutButtonStyleConfiguration
import com.klarna.mobile.sdk.api.checkout.express.KlarnaExpressCheckoutError
import com.klarna.mobile.sdk.api.checkout.express.KlarnaExpressCheckoutSessionOptions
import com.klarna.mobile.sdk.api.component.button.KlarnaButtonShape
import com.klarna.mobile.sdk.api.component.button.KlarnaButtonStyle
import com.klarna.mobile.sdk.api.component.button.KlarnaButtonTheme

class KlarnaExpressCheckoutButtonView(private val reactContext: ThemedReactContext) :
    FrameLayout(reactContext) {

    var clientToken: String? = null
    var sessionData: String? = null
    var autoFinalize: Boolean = true
    var collectShippingAddress: Boolean = false
    var buttonLocale: String = "en-US"
    var environment: String = "production"
    var region: String = "NA"
    var theme: String = "dark"
    var buttonShape: String = "rounded_rect"
    var buttonStyleValue: String = "filled"
    var returnUrlValue: String? = null
    var loggingLevel: String = "off"

    private var klarnaButton: KlarnaExpressCheckoutButton? = null

    private val callback = object : KlarnaExpressCheckoutButtonCallback {
        override fun onAuthorized(
            view: KlarnaExpressCheckoutButton,
            response: KlarnaExpressCheckoutButtonAuthorizationResponse
        ) {
            val payload: WritableMap = Arguments.createMap().apply {
                putBoolean("approved", response.approved)
                putBoolean("showForm", response.showForm)
                putBoolean("finalizeRequired", response.finalizeRequired)
                response.authorizationToken?.let { putString("authorizationToken", it) }
                response.clientToken?.let { putString("clientToken", it) }
                response.sessionId?.let { putString("sessionId", it) }
                response.collectedShippingAddress?.let { putString("collectedShippingAddress", it) }
                response.merchantReference1?.let { putString("merchantReference1", it) }
                response.merchantReference2?.let { putString("merchantReference2", it) }
            }
            dispatchEvent("onAuthorized", payload)
        }

        override fun onError(view: KlarnaExpressCheckoutButton, error: KlarnaExpressCheckoutError) {
            val payload: WritableMap = Arguments.createMap().apply {
                putString("name", error.name)
                putString("message", error.message)
                putBoolean("isFatal", error.isFatal)
                error.sessionId?.let { putString("sessionId", it) }
            }
            dispatchEvent("onError", payload)
        }
    }

    fun renderButton() {
        Handler(Looper.getMainLooper()).post {
            val token = clientToken
            if (token.isNullOrBlank()) {
                return@post
            }

            removeExistingButton()

            val sessionOptions = KlarnaExpressCheckoutSessionOptions.ServerSideSession(
                clientToken = token,
                sessionData = sessionData,
                autoFinalize = autoFinalize,
                collectShippingAddress = collectShippingAddress
            )

            val styleConfiguration = KlarnaExpressCheckoutButtonStyleConfiguration(
                theme = mapButtonTheme(theme),
                shape = mapButtonShape(buttonShape),
                style = mapButtonStyle(buttonStyleValue)
            )

            val options = KlarnaExpressCheckoutButtonOptions(
                sessionOptions = sessionOptions,
                callback = callback,
                locale = buttonLocale,
                styleConfiguration = styleConfiguration,
                environment = mapEnvironment(environment, region),
                region = mapRegion(region),
                theme = mapFlowTheme(theme),
                loggingLevel = mapLoggingLevel(loggingLevel)
            )

            val hostContext: Context = reactContext.currentActivity ?: reactContext
            klarnaButton = KlarnaExpressCheckoutButton(hostContext, options).also { btn ->
                btn.layoutParams = ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT
                )
                addView(btn)
            }
        }
    }

    private fun removeExistingButton() {
        klarnaButton?.let { removeView(it) }
        klarnaButton = null
    }

    private fun dispatchEvent(name: String, payload: WritableMap) {
        reactContext
            .getJSModule(RCTEventEmitter::class.java)
            .receiveEvent(id, name, payload)
    }

    private fun mapButtonTheme(value: String): KlarnaButtonTheme = when (value.lowercase()) {
        "light" -> KlarnaButtonTheme.LIGHT
        "auto" -> KlarnaButtonTheme.AUTO
        else -> KlarnaButtonTheme.DARK
    }

    private fun mapButtonShape(value: String): KlarnaButtonShape = when (value.lowercase()) {
        "rectangle" -> KlarnaButtonShape.RECTANGLE
        "pill" -> KlarnaButtonShape.PILL
        else -> KlarnaButtonShape.ROUNDED_RECT
    }

    private fun mapButtonStyle(value: String): KlarnaButtonStyle = when (value.lowercase()) {
        "outlined" -> KlarnaButtonStyle.OUTLINED
        else -> KlarnaButtonStyle.FILLED
    }

    private fun mapFlowTheme(value: String): KlarnaTheme = when (value.lowercase()) {
        "dark" -> KlarnaTheme.DARK
        "auto" -> KlarnaTheme.AUTOMATIC
        else -> KlarnaTheme.LIGHT
    }

    private fun mapEnvironment(env: String, region: String): KlarnaEnvironment {
        val upperRegion = region.uppercase()
        return when (env.lowercase()) {
            "playground" -> when (upperRegion) {
                "EU" -> KlarnaEnvironment.PLAYGROUND_EU
                "OC" -> KlarnaEnvironment.PLAYGROUND_OC
                else -> KlarnaEnvironment.PLAYGROUND_NA
            }
            else -> when (upperRegion) {
                "EU" -> KlarnaEnvironment.PRODUCTION_EU
                "OC" -> KlarnaEnvironment.PRODUCTION_OC
                else -> KlarnaEnvironment.PRODUCTION_NA
            }
        }
    }

    private fun mapRegion(value: String): KlarnaRegion = when (value.uppercase()) {
        "EU" -> KlarnaRegion.EU
        "OC" -> KlarnaRegion.OC
        else -> KlarnaRegion.NA
    }

    private fun mapLoggingLevel(value: String): KlarnaLoggingLevel = when (value.lowercase()) {
        "error" -> KlarnaLoggingLevel.Error
        "verbose" -> KlarnaLoggingLevel.Verbose
        else -> KlarnaLoggingLevel.Off
    }
}

File: packages/@juspay-tech/react-native-hyperswitch-klarna-expresscheckout/ios/KlarnaExpressCheckoutButtonView.swift
import Foundation
import UIKit
import React
import KlarnaMobileSDK

@objc(KlarnaExpressCheckoutButtonView)
class KlarnaExpressCheckoutButtonView: UIView {

  @objc var onAuthorized: RCTDirectEventBlock?
  @objc var onError: RCTDirectEventBlock?

  @objc dynamic var clientToken: NSString = "" {
    didSet { rebuildButton() }
  }
  @objc dynamic var sessionData: NSString = "" {
    didSet { rebuildButton() }
  }
  @objc dynamic var autoFinalize: Bool = true {
    didSet { rebuildButton() }
  }
  @objc dynamic var collectShippingAddress: Bool = false {
    didSet { rebuildButton() }
  }
  @objc dynamic var locale: NSString = "en-US" {
    didSet { rebuildButton() }
  }
  @objc dynamic var environment: NSString = "production" {
    didSet { rebuildButton() }
  }
  @objc dynamic var region: NSString = "NA" {
    didSet { rebuildButton() }
  }
  @objc dynamic var theme: NSString = "dark" {
    didSet { rebuildButton() }
  }
  @objc dynamic var buttonShape: NSString = "rounded_rect" {
    didSet { rebuildButton() }
  }
  @objc dynamic var buttonStyleValue: NSString = "filled" {
    didSet { rebuildButton() }
  }
  @objc dynamic var returnUrl: NSString = "" {
    didSet { rebuildButton() }
  }
  @objc dynamic var loggingLevel: NSString = "off" {
    didSet { rebuildButton() }
  }

  private var klarnaButton: KlarnaExpressCheckoutButton?

  override init(frame: CGRect) {
    super.init(frame: frame)
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    if window != nil {
      rebuildButton()
    }
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    klarnaButton?.frame = bounds
  }

  private func rebuildButton() {
    DispatchQueue.main.async { [weak self] in
      guard let self = self else { return }

      self.klarnaButton?.removeFromSuperview()
      self.klarnaButton = nil

      let token = self.clientToken as String
      guard !token.isEmpty else { return }

      let sessionOptions = KlarnaExpressCheckoutSessionOptions.ServerSideSession(
        clientToken: token,
        sessionData: (self.sessionData as String).isEmpty ? nil : (self.sessionData as String),
        autoFinalize: self.autoFinalize,
        collectShippingAddress: self.collectShippingAddress
      )

      let styleConfiguration = KlarnaExpressCheckoutButtonStyleConfiguration(
        theme: self.mapButtonTheme(self.theme as String),
        shape: self.mapButtonShape(self.buttonShape as String),
        style: self.mapButtonStyle(self.buttonStyleValue as String)
      )

      let returnUrlString = self.returnUrl as String
      let effectiveReturnUrl = returnUrlString.isEmpty ? "hyperswitchsdk://klarna" : returnUrlString

      let options = KlarnaExpressCheckoutButtonOptions(
        sessionOptions: sessionOptions,
        returnUrl: effectiveReturnUrl,
        delegate: self,
        locale: self.locale as String,
        styleConfiguration: styleConfiguration,
        environment: self.mapEnvironment(self.environment as String),
        region: self.mapRegion(self.region as String),
        theme: self.mapFlowTheme(self.theme as String),
        loggingLevel: self.mapLoggingLevel(self.loggingLevel as String)
      )

      let button = KlarnaExpressCheckoutButton(options: options)
      button.frame = self.bounds
      button.autoresizingMask = [.flexibleWidth, .flexibleHeight]
      self.addSubview(button)
      self.klarnaButton = button
    }
  }

  private func mapButtonTheme(_ value: String) -> KlarnaButtonTheme {
    switch value.lowercased() {
    case "light": return .light
    case "auto": return .auto
    default: return .dark
    }
  }

  private func mapButtonShape(_ value: String) -> KlarnaButtonShape {
    switch value.lowercased() {
    case "rectangle": return .rectangle
    case "pill": return .pill
    default: return .rounded_rect
    }
  }

  private func mapButtonStyle(_ value: String) -> KlarnaButtonStyle {
    switch value.lowercased() {
    case "outlined": return .outlined
    default: return .filled
    }
  }

  private func mapFlowTheme(_ value: String) -> KlarnaTheme {
    switch value.lowercased() {
    case "dark": return .dark
    case "auto": return .automatic
    default: return .light
    }
  }

  private func mapEnvironment(_ value: String) -> KlarnaEnvironment {
    switch value.lowercased() {
    case "playground": return .playground
    default: return .production
    }
  }

  private func mapRegion(_ value: String) -> KlarnaRegion {
    switch value.uppercased() {
    case "EU": return .eu
    case "OC": return .oc
    default: return .na
    }
  }

  private func mapLoggingLevel(_ value: String) -> KlarnaLoggingLevel {
    switch value.lowercased() {
    case "error": return .error
    case "verbose": return .verbose
    default: return .off
    }
  }
}

extension KlarnaExpressCheckoutButtonView: KlarnaExpressCheckoutButtonDelegate {
  func onAuthorized(
    view: KlarnaExpressCheckoutButton,
    response: KlarnaExpressCheckoutButtonAuthorizationResponse
  ) {
    var payload: [String: Any] = [
      "approved": response.approved,
      "showForm": response.showForm,
      "finalizeRequired": response.finalizeRequired,
    ]
    if let token = response.authorizationToken { payload["authorizationToken"] = token }
    if let clientToken = response.clientToken { payload["clientToken"] = clientToken }
    payload["sessionId"] = response.sessionId
    if let shipping = response.collectedShippingAddress {
      payload["collectedShippingAddress"] = shipping
    }
    if let ref1 = response.merchantReference1 { payload["merchantReference1"] = ref1 }
    if let ref2 = response.merchantReference2 { payload["merchantReference2"] = ref2 }

    onAuthorized?(payload)
  }

  func onError(view: KlarnaExpressCheckoutButton, error: KlarnaError) {
    var payload: [String: Any] = [
      "name": error.name,
      "message": error.message,
      "isFatal": error.isFatal,
    ]
    if let sid = error.sessionId { payload["sessionId"] = sid }
    onError?(payload)
  }
}
